src/models/batch_auction_model.rs

```diff
@@ -160,6 +170,155 @@ impl SettledBatchAuctionModel {
+    }
+
+    pub fn insert_new_price(
+        &mut self,
+        splitted_trade_amounts: &HashMap<(H160, H160), (U256, U256)>,
+        query: SwapQuery,
+        swap: SwapResponse,
+        tokens: &BTreeMap<H160, TokenInfoModel>,
+    ) -> Result<()> {
+        let src_token = query.sell_token;
+        let dest_token = query.buy_token;
+        let (sell_amount, buy_amount) = match (
+            splitted_trade_amounts.get(&(src_token, dest_token)),
+            splitted_trade_amounts.get(&(dest_token, src_token)),
+        ) {
+            (Some((_sell_amount, _)), Some((buy_amount, substracted_sell_amount))) => {
+                (*substracted_sell_amount, *buy_amount)
+            }
+            (Some((_, _)), None) => (U256::zero(), U256::zero()),
+            (None, Some((_, _))) => (U256::zero(), U256::zero()),
+            (None, None) => (U256::zero(), U256::zero()),
+        };
+        let (sell_amount, buy_amount) = (
+            sell_amount.checked_add(swap.sell_amount).unwrap(),
+            buy_amount.checked_add(swap.buy_amount).unwrap(),
+        );
+
+        match (
+            self.prices.clone().get(&src_token),
+            self.prices.clone().get(&dest_token),
+        ) {
+            (Some(_), Some(_)) => return Err(anyhow!("can't deal with such a ring")),
+            (Some(price_sell_token), None) => {
+                self.prices.insert(
+                    query.buy_token,
+                    price_sell_token
+                        .checked_mul(sell_amount)
+                        .unwrap()
+                        .checked_div(buy_amount)
+                        .unwrap(),
+                );
+            }
+            (None, Some(price_buy_token)) => {
+                self.prices.insert(
+                    query.sell_token,
+                    price_buy_token
+                        .checked_mul(buy_amount)
+                        .unwrap()
+                        .checked_div(sell_amount)
+                        .unwrap(),
+                );
+            }
+            (None, None) => {
+                if self.prices.is_empty() {
+                    self.prices.insert(
+                        query.sell_token,
+                        buy_amount.checked_mul(U256::from(SCALING_FACTOR)).unwrap(),
+                    );
+                    self.prices.insert(
+                        query.buy_token,
+                        sell_amount.checked_mul(U256::from(SCALING_FACTOR)).unwrap(),
+                    );
+                } else {
+                    // If there are independent trades, e.g. DAI -> USDC AND ETH -> GNO, the prices between
+                    // DAI and GNO still need to satisfy the price check in the driver. Hence, the prices of unrelated trades
+                    // still needs to consider the external prices provided for the auction for the setting of the new price
+                    for (token, token_price) in self.prices.iter() {
```

Comment:
Is the for-loop here needed just because some tokens might not have external price? I am asking since there is a `break` once there is a match.

Reply:
>Is the for-loop here needed just because some tokens might not have external price? I am asking since there is a break once there is a match.

Yes, exactly

>Also, as you are computing more and more prices, is it obvious that you are getting consistent prices here? What I mean is that is it clear that the newly added prices are "compatible w.r.t to the test" with all previously computed prices?

I see it like this: If the previous price solution is price compatible and then I add a new pair with this, then the added pair should be price compatible, as it is added with the right ratio. 
But yeah, the adding is not really transitive: If I am adding A-> B and D->C , but A-D has a price variation, then B->C will also have a price variation and at some point these price variations could be too big. 
But I don't care about this effect, as its quite uncommon to have completely unrelated pairs.